proxy/proxy/httpserver.py

```diff
@@ -0,0 +1,171 @@
+"""proxy server"""
+from http import HTTPStatus
+from http.server import ThreadingHTTPServer, BaseHTTPRequestHandler
+import json
+import logging
+import os
+import ssl
+import zlib
+import requests
+
+logging.basicConfig(level=logging.DEBUG)
+
+HOST = "127.0.0.1"
+PORT = 8443
+TIMEOUT = 30
+
+TEST_DST_PROTOCOL = "http"
+TEST_GATEWAY_URL = "127.0.0.1:60000"
+dst_protocol = os.environ.get("DST_PROTOCOL", TEST_DST_PROTOCOL)
+gateway_url = os.environ.get("GATEWAY_URL", TEST_GATEWAY_URL)
+
+
+class ProxyRequestHandler(BaseHTTPRequestHandler):
+    """Proxy Request Handler."""
+
+    def gzip_encode(self, content):
+        """gzip encode"""
+        gzip_compress = zlib.compressobj(9, zlib.DEFLATED, zlib.MAX_WBITS | 16)
+        data = gzip_compress.compress(content) + gzip_compress.flush()
+        return data
+
+    def handle_response(self, resp):
+        """handle response."""
+        self.send_response(resp.status_code)
+        if resp.content and len(resp.content) != 0:
+            content = resp.content
+            logging.debug("Recieved response content: %s", content.decode("utf-8"))
+            if (
+                "content-encoding" in resp.headers
+                and resp.headers["content-encoding"] == "gzip"
+            ):
+                content = self.gzip_encode(content)
+            if (
+                "Transfer-Encoding" in resp.headers
+                and resp.headers["Transfer-Encoding"] == "chunked"
+            ):
+                content = (
+                    bytes(f"{len(content):x}", "utf-8")
+                    + b"\x0d\x0a"
+                    + content
+                    + b"\x0d\x0a0\x0d\x0a\x0d\x0a"
+                )
+        for header in resp.headers:
+            self.send_header(header, resp.headers[header])
```

Comment:
I don't know if this can be used here but it seems `http.server` has methods to modify `send_headers` and `end_headers`:

https://stackoverflow.com/questions/12499171/can-i-set-a-header-with-pythons-simplehttpserver

Reply:
I believe that It has `end_headers` but doesn't have `send_headers`. https://github.com/python/cpython/blob/3.12/Lib/http/server.py

Reply:
Thanks for the link, Aki. I will take a look 🙏